internal/provider/data_source_vault_secrets_app.go

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"log"
+
+	sharedmodels "github.com/hashicorp/hcp-sdk-go/clients/cloud-shared/v1/models"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/hashicorp/terraform-provider-hcp/internal/clients"
+)
+
+func dataSourceVaultSecretsApp() *schema.Resource {
+	return &schema.Resource{
+		Description: "The Vault Secrets app data source retrieves secrets and their latest version values for a given application.",
+		ReadContext: dataSourceVaultSecretsAppRead,
+		Schema: map[string]*schema.Schema{
+			"app_name": {
+				Description:      "The name of the Vault Secrets application.",
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateDiagFunc: validateSlugID,
+			},
+			"organization_id": {
+				Description: "The ID of the HCP organization where the Vault Secrets app is located.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"project_id": {
+				Description: "The ID of the HCP project where the Vault Secrets app is located.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"secrets": {
+				Description: "A map of all secrets in the Vault Secrets app. Key is the secret name, value is the latest secret version value.",
+				Type:        schema.TypeMap,
+				Computed:    true,
+				Sensitive:   true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func dataSourceVaultSecretsAppRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	appName, ok := d.Get("app_name").(string)
+	if !ok {
+		return diag.Errorf("Failed to read app_name during data source Read.")
+	}
+	client := meta.(*clients.Client)
+	loc := &sharedmodels.HashicorpCloudLocationLocation{
+		OrganizationID: client.Config.OrganizationID,
+		ProjectID:      client.Config.ProjectID,
```

Comment:
The provider now has first-class support for multi-project, which includes a pattern where individual resources should support an Optional `project_id` attribute. This new data source should probably also support that pattern.